// react-outer-click/lib/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime/helpers/extends');
var _objectWithoutPropertiesLoose = require('@babel/runtime/helpers/objectWithoutPropertiesLoose');
var PropTypes = require('prop-types');
var React = require('react');
var useComposedRef = require('use-composed-ref');
var invariant = require('tiny-invariant');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return Object.freeze(n);
}

var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _objectWithoutPropertiesLoose__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutPropertiesLoose);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var React__namespace = /*#__PURE__*/_interopNamespace(React);
var useComposedRef__default = /*#__PURE__*/_interopDefaultLegacy(useComposedRef);
var invariant__default = /*#__PURE__*/_interopDefaultLegacy(invariant);

var castArray = function castArray(value) {
  return Array.isArray(value) ? value : [value];
};

var useOuterClick = function useOuterClick(refs, handler) {
  !(Array.isArray(refs) || typeof refs === 'object' && refs !== null) ? process.env.NODE_ENV !== "production" ? invariant__default['default'](false, 'Expected `refs` to be an array or object') : invariant__default['default'](false) : void 0;

  if (Array.isArray(refs)) {
    !(refs.length > 0) ? process.env.NODE_ENV !== "production" ? invariant__default['default'](false, 'Expected `refs` to to not be empty') : invariant__default['default'](false) : void 0;
    refs.forEach(function (ref, i) {
      !(typeof ref === 'object' && ref !== null) ? process.env.NODE_ENV !== "production" ? invariant__default['default'](false, "Expected `refs[" + i + "]` to be an object") : invariant__default['default'](false) : void 0;
    });
  }

  !(typeof handler === 'function' || handler == null) ? process.env.NODE_ENV !== "production" ? invariant__default['default'](false, 'Expected `handler` to be a function') : invariant__default['default'](false) : void 0;
  var eventListenerRef = React.useRef();
  React.useEffect(function () {
    eventListenerRef.current = function (event) {
      if (castArray(refs).every(function (ref) {
        return !ref.current || ref.current !== event.target && !ref.current.contains(event.target);
      })) {
        handler == null ? void 0 : handler(event);
      }
    };
  }, [handler, refs]);
  React.useEffect(function () {
    var eventListener = function eventListener(event) {
      eventListenerRef.current(event);
    };

    ['mousedown', 'touchstart'].forEach(function (type) {
      document.addEventListener(type, eventListener, true);
    });
    return function () {
      ['mousedown', 'touchstart'].forEach(function (type) {
        document.removeEventListener(type, eventListener, true);
      });
    };
  }, []);
};

var OuterClick = /*#__PURE__*/React__namespace.forwardRef(function OuterClick(_ref, userRef) {
  var _ref$as = _ref.as,
      Element = _ref$as === void 0 ? 'div' : _ref$as,
      _ref$children = _ref.children,
      children = _ref$children === void 0 ? null : _ref$children,
      _ref$onOuterClick = _ref.onOuterClick,
      onOuterClick = _ref$onOuterClick === void 0 ? null : _ref$onOuterClick,
      rest = _objectWithoutPropertiesLoose__default['default'](_ref, ["as", "children", "onOuterClick"]);

  var libRef = React.useRef();
  var ref = useComposedRef__default['default'](libRef, userRef);
  useOuterClick(libRef, onOuterClick);
  return /*#__PURE__*/React__namespace.createElement(Element, _extends__default['default']({}, rest, {
    ref: ref
  }), children);
});
OuterClick.propTypes = process.env.NODE_ENV !== "production" ? {
  as: PropTypes__default['default'].elementType,
  children: PropTypes__default['default'].node,
  onOuterClick: PropTypes__default['default'].func
} : {};

exports.OuterClick = OuterClick;
exports.useOuterClick = useOuterClick;
